### ISSUE TITLE
Develop a formulation for better handling of adjoint cross sections

### ISSUE THREAD

**Issue by zachhardy** [chi-tech/chi-tech#313]
Currently, the adjoint solver lugs around the transpose of the transfer and production matrices within `matid_to_S_transpose` and `matid_to_F_transpose`. This is done instead of flipping indices for performance reasons. Ideally, there is some more natural accessor, such as a derived `TransportCrossSections` class `AdjointTransportCrossSections`, containing the necessary transposed data, or the cross sections are modified within `TransportCrossSections` with a flag to determine whether or not modifications have been made. The favored method was to leave `TransportCrossSections` alone.

**Comment by ragusa** [comment]
When one loads a XS in chi-tech, one has an object XSobj_forward.

Can we create a derived class that 
- takes in XSobj_forward, make it adjoint, which can mean transposing but maybe also reversing group ordering (**think about that one; it may make re-using LBS easier**)
- and returns to lua XSobj_adjoint. Then, it is the user resposability to pass the proper lua object tot he adjoint solver.

@Naktakala @zachhardy Thoughts?

**Comment by zachhardy** [comment]
@ragusa This is what I am thinking. We can create an object `AdjointTransportCrossSections`. Then we have two options:

1. Make this a derived type of `TransportCrossSections` and use a few initialization methods. First, we can initialize via an instance of `TransportCrossSections` where we copy all vector cross section data and transpose the transfer and production matrices. Second, we can initialize via a .cxs file where the data will be transposed when read. Alternatively, we can have an adjoint .cxs file read in which has the data pre-transposed. 
2. We can create a struct `AdjointTransportCrossSections` which stores only the transposed transfer and production matrices. This can be initialized in the same ways as in the previous item. From this, we avoid copies of all of the other cross section data. During the `SetSource` routine we then query the forward cross sections and the adjoint cross sections.

**Comment by ragusa** [comment]
@zachhardy 
Also, do not overlook the possiblity of transposing  AND reversing the groups (last row becomes first row and so on). Imagine you have a pure down scatterer. If you transpose, the transfer matrix is now upper triangular. It would best to solver from the most thermal group upwards to the first fast group. 

- Maybe this reversing of the group solve is obvious from lua. 
- But even if so, you need to think about the WGS and AGS or the group sets in general. I just have a feeling that it may be worthwhile to also reverse so that the loops so that LBS remains as before, more or less.

**Comment by zachhardy** [comment]
@ragusa 
That is a good point. Currently the groups are ordered normally and the transpose matrices are queried. This is synonymous with option 2 above where the queried transpose matrices simply belong to a struct. Your argument bolsters the idea of using a derived type instead of a standalone struct with the transposed data.

**Comment by jemorel** [comment]
If you reverse the groups, don't forget to restore the original order to the solution, otherwise it will mess up the inner-products for the response (unless you build that adjoint ordering into your inner-product routines.)

--Jim
*********************************************
Jim  E. Morel, Ph.D.
Professor
Director, Center for Large Scale Scientific Simulations
Director, Center for Exascale Radiation Transport
Department of Nuclear Engineering, TAMU 3133
Texas A&M University
College Station, TX 77843
Phone: (979) 845-6072
Cell: (979) 204-8572
Fax: (979) 845-6443
********************************************
________________________________

**Comment by ragusa** [comment]
This is correct. The QoI using the adjoint flux likely have some forward function, so the ordering of groups need to be consistent. At the end of the adjoint solve, I would simply re-order of the adjoint flux moments to be back in the standard forward ordering.

A small correction: if you reverse the adjoint flux ordering, the transfert matrix (already transposed to get the adjoint) also needs to have a reversing of the groups as well (of course), but this implies reversing both the rows and the columns of the transfer matrix (and this is not the same as transposing; it is: A_{i,j} becomes A_{G-i, G-j})

**Comment by Naktakala** [comment]
## On the topic of inverting the group order: 
This might seem like it would enable a lot of code reuse but it really doesn't. I've certainly tried it but it makes visualization a mess (group 68 of 175 adjoint grps is actually what in the forward problem?), it requires a mapping when writing the flux moments (.data files) for reuse during a separate inner product, mapping when writing restart files, and is generally VERY confusing to read in the code, especially for evaluating responses where one can benefit from having everything consistent. Where the thermal groups start or end is also inverted then.

Since we do Jacobi iteration on the groups **within** a groupset, it is only necessary to invert the groupset-order. This is easily done in the input language. In fact, with the new AGS solver scheme you dont even have to because you can just add across groupset iterations. Thereafter... Our Set-Source routine changes, i.e., remove external sources, remove point sources, add volumetric response sources (QOIs). The surface response source is present during the sweep chunk but should be catered for with boundary conditions.

We can still play with options on the `SetSource` routine.

## On the topic of another XS class
I don't believe we need to physically make another set of XSs in the input code. That would overwhelm a user trying to keep track of all the XSs he/she/it already created and add unnecessary input language. I believe there is one good solution to this: In the same sense as one would do a "matrix transpose by proxy", we do a `AdjointXS` by proxy.  It goes like this
```c++
//Absolute abstract class
class XSBase
{
public:
  virtual const std::vector<double>& sigma_t() const = 0;
  virtual const chi_math::SparseMatrix& transfer_matrix(int ell) const = 0;
  virtual const chi_math::DenseMatrix& production_matrix() const = 0;
};

//Equivalent to what we have now
class TransportXS : public XSBase
{
protected:
  std::vector<double> sigma_t_;
  std::vector<chi_math::SparseMatrix> transfer_matrices_;
  chi_math::DenseMatrix production_matrix_;
public:
  //Some constructors for reading from file or 
  //manually setting
  
  const std::vector<double>& sigma_t() const override 
  { return sigma_t_; }
  const chi_math::SparseMatrix& transfer_matrix(int ell) const override 
  { return transfer_matrix_.at(ell); }
  const chi_math::DenseMatrix& production_matrix() const override
  { return production_matrix_; }
};

//Adjoint by proxy
class AdjointXS : public TransportXS
{
protected:
  const TransportXS& reference_xs_;
  std::vector<chi_math::SparseMatrix> transposed_transfer_matrices_;
  chi_math::DenseMatrix transposed_production_matrix_;
public:
  //Constructor
  AdjointXS(TransportXS& reference_xs) : reference_xs_(reference_xs) 
  {
    //Compute transposed_transfer_matrices_
    //Compute transposed_production_matrix_
  }

  const std::vector<double>& sigma_t() const override 
  { return reference_xs_.sigma_t(); }
  const chi_math::SparseMatrix& transfer_matrix(int ell) const override 
  { return transposed_transfer_matrices_.at(ell); }
  const chi_math::DenseMatrix& production_matrix() const override
  { return transposed_production_matrix_; }
}
```

Of course, I left out some of the other data members but at least we don't have to copy ALL the data. We could technically have calls to the matrices, i.e., `TransferMatrix(int ell, int g, int gprime)` then override by transposing the `g` and `grpime` indices, however, we gain a lot of speed by using iterators so it will probably be worth it to just save the transposed operators.

The adjoint solver then initializes these proxy XSs that it can use. We also refactor all the LBS machinery to always operate on the `XSBase` to provide a level of abstraction.

**Comment by Naktakala** [comment]
Also a good time to rename `TransportCrossSections` to `MultigroupXSs`.

**Comment by ragusa** [comment]
YES. The first time I read this variable name, I thought of transport-corrected XS and it confused me!
